From us15@os.inf.tu-dresden.de Wed Jun  6 14:34:18 2007
Return-Path: <us15@os.inf.tu-dresden.de>
X-Spam-Checker-Version: SpamAssassin 3.1.7-deb (2006-10-05) on debian
X-Spam-Level: 
X-Spam-Status: No, score=0.1 required=5.0 tests=AWL,MAILTO_TO_SPAM_ADDR 
	autolearn=no version=3.1.7-deb
Received: from os.inf.tu-dresden.de (os.inf.tu-dresden.de [141.76.48.99])
	(using TLSv1 with cipher DHE-RSA-AES256-SHA (256/256 bits)) (No client
	certificate requested) by mail.tglx.de (Postfix) with ESMTP id CB67965C065
	for <tglx@linutronix.de>; Wed,  6 Jun 2007 14:34:18 +0200 (CEST)
Received: from nova.inf.tu-dresden.de ([141.76.48.73]
	helo=laptop.hypervisor.org) by os.inf.tu-dresden.de with esmtpsa
	(TLSv1:AES256-SHA:256) (Exim 4.67) id 1HvuiQ-0000WF-8q; Wed, 06 Jun 2007
	14:34:18 +0200
Date: Wed, 6 Jun 2007 14:34:14 +0200
From: "Udo A. Steinberg" <us15@os.inf.tu-dresden.de>
To: Thomas Gleixner <tglx@linutronix.de>, Venkatesh Pallipadi <venkatesh.pallipadi@intel.com>
Subject: [PATCH]: Enable HPET on ICH3 and ICH4
Message-ID: <20070606143414.6003edd0@laptop.hypervisor.org>
X-Mailer: X-Mailer 5.0 Gold
Mime-Version: 1.0
Content-Type: multipart/signed; boundary=Sig_TyoZ8hpf907DzN6.B9sCrGr; protocol="application/pgp-signature"; micalg=PGP-SHA1
X-Evolution-Source: imap://tglx%40linutronix.de@localhost:8993/

ICH3 and ICH4 have undocumented HPET capabilities. This patch enables HPET
for platforms based around these ICHs. 

Tested on various ICH3 and ICH4 platforms.

Because HPET is not officially documented for ICH3/4 and may not have been
validated by chipset folks, we're on thin ice here. I'd recommend testing 
this patch in -hrt or -mm for a while and wait for success/failure reports
before feeding it upstream.

Signed-off-by: Udo A. Steinberg <us15@os.inf.tu-dresden.de>

---
 arch/i386/kernel/quirks.c |    8 ++++++++
 1 file changed, 8 insertions(+)

Index: linux-rt.q/arch/i386/kernel/quirks.c
===================================================================
--- linux-rt.q.orig/arch/i386/kernel/quirks.c
+++ linux-rt.q/arch/i386/kernel/quirks.c
@@ -232,6 +232,14 @@ static void old_ich_force_enable_hpet(st
 	printk(KERN_DEBUG "Failed to force enable HPET\n");
 }
 
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801CA_0,
+                         old_ich_force_enable_hpet);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801CA_12,
+                         old_ich_force_enable_hpet);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801DB_0,
+                         old_ich_force_enable_hpet);
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801DB_12,
+                         old_ich_force_enable_hpet);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801EB_0,
                          old_ich_force_enable_hpet);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_INTEL, PCI_DEVICE_ID_INTEL_82801EB_12,
